solutions/cses-1625.mdx

```diff
@@ -0,0 +1,281 @@
+
+• running time: 1.8 seconds
+
+• number of recursive calls: 221 million
+
+### Optimization 3
+
+The idea of Optimization 3 can be generalized: if the path cannot continue
+forward but can turn either left or right, the grid splits into two parts that
+both contain unvisited squares. It is clear that we cannot visit all squares
+anymore, so we can terminate the search.
+
+• running time: 0.6 seconds
+
+• number of recursive calls: 69 million
+
+Now is a good moment to stop optimizing the algorithm and see what we have
+achieved. The running time of the original algorithm was 483 seconds, and now
+after the optimizations, the running time is only 0.6 seconds. Thus, the
+algorithm became nearly 1000 times faster after the optimizations.
+
+<Info>
+This is a usual phenomenon in backtracking, because the search tree is usually
+large and even simple observations can effectively prune the search. Especially
+useful are optimizations that occur during the first steps of the algorithm, i.e., at
+the top of the search tree.
+</Info>
+
+For C++ users, this will be enough pruning to make the program run in time under
+worst case conditions. However, for Java and Python users, this is still too
+slow (Java takes around 1.2 seconds for the worst case). Thus, we will need to
+further optimize our search.
+
+### Optimization 4
+
+If the path creates a dead end that is not the bottom left corner, either the
+path will fail to visit all squares (the path may stop at the dead end or pass
+over it, sealing a square off) or the path will end in the wrong location. Thus
+we want to avoid creating dead ends. For example, if the square to the left of
+our current location is blocked on three sides (including our current location),
+then the next step must be to the left in order to avoid creating a dead end.
+After this optimization, the program runs in under 1 second.
+
+## Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <iostream>
+using namespace std;
+
+const int dirLen = 4;
+int dr[dirLen] = {-1, 0, 1, 0};
+int dc[dirLen] = {0, 1, 0, -1};
+const int pathLen = 48;
+int p[pathLen];
+const int gridSize = 9;
+bool onPath[gridSize][gridSize];
+
+int tryPath(int pathIdx, int curR, int curC) {
+	if ((onPath[curR][curC - 1] && onPath[curR][curC + 1])
+			&& (!onPath[curR - 1][curC] && !onPath[curR + 1][curC])) return 0;
+	if ((onPath[curR - 1][curC] && onPath[curR + 1][curC])
+					&& (!onPath[curR][curC - 1] && !onPath[curR][curC + 1])) return 0;
+
+	if (curR == 7 && curC == 1) { // reached endpoint before visiting all
+		if (pathIdx == 48) return 1;
+		return 0;
+	}
+
+	if (pathIdx == 48) return 0;
+
+	int ret = 0;
+	onPath[curR][curC] = true;
+
+	// turn already determined:
+	if (p[pathIdx] < 4) {
+		int nxtR = curR + dr[p[pathIdx]];
+		int nxtC = curC + dc[p[pathIdx]];
+		if (!onPath[nxtR][nxtC]) ret += tryPath(pathIdx + 1, nxtR, nxtC);
+	} 
+  // see Java solution for optimization 4 implementation
```

Comment:
indentation seems hecked up

Reply:
1. Thank you for not swearing.
2. Is it better now? I couldn't tell which lines in specific you were pointing to.

Reply:
no, the indentation is still wrong on line 128

Reply:
Indentation issues for all comments _should_ be fixed now?